File: CodingInterviewsPythonSolutions/Longest Valid Parentheses.py
class Solution:
    def longestValidParentheses(self, s: str) -> int:
        sz = len(s)
        result = 0
        stk = [-1]

        for i in range(sz):
            if s[i] == '(':
                stk.append(i)
            else:
                stk.pop()

                if not stk:
                    stk.append(i)
                
                result = max(result, i - stk[-1])

        return result

File: CodingInterviewsJavaSolutions/Longest Valid Parentheses.java
class Solution {
    public int longestValidParentheses(String s) {
        int sz = s.length();
        int result = 0;
        Stack<Integer> stk = new Stack<>();
        stk.push(-1);

        for (int i = 0; i < sz; i++) {
            if (s.charAt(i) == '(')
                stk.push(i);
            else {
                stk.pop();

                if (stk.empty())
                    stk.push(i);

                result = Math.max(result, i - stk.peek());
            }
        }
        return result;
    }
}
